### ISSUE TITLE
Nothing tells you when you're timing out on 30 second harvester proof checks

### ISSUE THREAD

**Issue by coding-horror** [Chia-Network/chia-blockchain#2651]
**The problem**

I thought I was farming, but I wasn't -- because something about my network caused the proof check to take more than the hard-coded 30 second limit.

I had an average time to win of 8 or 9 hours for more than 120 hours without a single win. This seemed statistically implausible, so I researched the logs, and cleared any errors or warnings in the logs (well done, all the warnings and errors in `debug.log` were [indeed things I should fix!](https://chiaforum.com/t/protip-check-your-debug-log-regularly/411)). Still no wins for a long time.

**How to reproduce**

Have a bunch of plots on slow storage media; when the proof check happens, verifying the proofs takes longer than the hard-coded 30 seconds allowed. You will never win a single Chia, but there's absolutely nothing in the GUI to inform you that this is happening. You can view the logs, but in the logs it is not even presented as a warning (!), but as an INFO message:

```
2021-04-24T16:03:29.039 harvester chia.harvester.harvester: INFO     8 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 3.74567 s. Total 3922 plots
2021-04-24T16:03:29.433 harvester chia.harvester.harvester: INFO     7 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 2.92976 s. Total 3922 plots
2021-04-24T16:03:29.635 harvester chia.harvester.harvester: INFO     7 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 2.01540 s. Total 3922 plots
2021-04-24T16:03:44.553 harvester chia.harvester.harvester: INFO     4 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 12.11716 s. Total 3922 plots
2021-04-24T16:04:10.882 harvester chia.harvester.harvester: INFO     6 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 24.51660 s. Total 3922 plots
2021-04-24T16:04:36.101 harvester chia.harvester.harvester: INFO     9 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 37.96922 s. Total 3922 plots
2021-04-24T16:05:13.959 harvester chia.harvester.harvester: INFO     11 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 63.07326 s. Total 3922 plots
2021-04-24T16:05:13.959 harvester chia.harvester.harvester: INFO     5 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 50.81979 s. Total 3922 plots
2021-04-24T16:05:31.022 harvester chia.harvester.harvester: INFO     6 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 54.92095 s. Total 3922 plots
2021-04-24T16:05:31.022 harvester chia.harvester.harvester: INFO     7 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 42.29693 s. Total 3922 plots
2021-04-24T16:05:31.022 harvester chia.harvester.harvester: INFO     8 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 29.67779 s. Total 3922 plots
2021-04-24T16:05:33.568 harvester chia.harvester.harvester: INFO     4 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 18.59361 s. Total 3922 plots
2021-04-24T16:05:33.568 harvester chia.harvester.harvester: INFO     8 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 6.10934 s. Total 3922 plots
2021-04-24T16:05:33.568 harvester chia.harvester.harvester: INFO     10 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 3.17564 s. Total 3922 plots
2021-04-24T16:05:33.568 harvester chia.harvester.harvester: INFO     13 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 3.82808 s. Total 3922 plots
2021-04-24T16:05:35.600 harvester chia.harvester.harvester: INFO     5 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 2.82815 s. Total 3922 plots
2021-04-24T16:05:35.600 harvester chia.harvester.harvester: INFO     12 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 3.56252 s. Total 3922 plots
2021-04-24T16:05:58.022 harvester chia.harvester.harvester: INFO     5 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 12.26593 s. Total 3922 plots
2021-04-24T16:06:11.887 harvester chia.harvester.harvester: INFO     4 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 12.42098 s. Total 3922 plots
2021-04-24T16:06:36.417 harvester chia.harvester.harvester: INFO     3 plots were eligible for farming 1c75d8d21c... Found 0 proofs. Time: 24.36354 s. Total 3922 plots
```

Of the above, the proofs that take longer than 30 seconds are not eligible to win, but this is not logged as an ERROR or WARNING or surfaced in the UI in any way.

**Expected behavior**

The GUI will tell you "hey, your proof checks are too slow, there's absolutely _no_ chance for you to win, even if you are farming infinity plots"

**Screenshots**

![Imgur](https://imgur.com/zVWEWCP.png)

**Desktop**

- OS: Windows 10
- OS Version/Flavor: latest 19042.928
- CPU: i7-1165G7

**Additional context**

I followed up on the #support channel in Keybase, where I got the important advice to enable INFO level logging and check for the 30 second proof limit.. and [I wrote up a detailed account on the forum](https://chiaforum.com/t/troubleshooting-failure-to-farm-with-lots-of-plots-due-to-harvester-30s-timeout/413); if you need excruciating levels of detail, please check there 🙇‍♂️

**Recommended solution**

- Allow the 30 second proof check threshold to be configurable to account for slower storage (or relax it a bit)
- change the log level of beyond 30 second on harvester proof checks to WARN or ERROR
- in the logs, provide details on which files are being checked for proofs, so you can identify which media or NAS is potentially the problem one slowing everything down

**Comment by maomingyang1314** [comment]
I think you can assemble another computer and participate in farming

**Comment by maomingyang1314** [comment]
Can you attach a screenshot of your explorer? Let me take a look at which one has the problem

**Comment by maomingyang1314** [comment]
Moreover, I would like to know why your laptop can mount nearly 400TB disks, is your disk on nas?

**Comment by maomingyang1314** [comment]
Your NAS caused your task to time out. It is recommended to abandon this method.

**Comment by maomingyang1314** [comment]
This issue has been resolved, please close the issue

**Comment by AZ663** [comment]
多少速度会影响呢

**Comment by mariano54** [comment]
We can't change the consensus algorithm anymore, but we can change the log level, and show the files. 

Please note that looking up qualities for plots passing the filter requires about 7 random reads in a plot, whereas actually looking up a proof requires 64 reads. It might not be feasible on a slow NAS, since these are sequential reads. Furthermore, you need to take into account network latency to propagate your proof and block to the network, so you should be under 5 seconds to reduce risk of losing rewards. 


Actually, the proof of space library does them sequentially, but they could be done in parallel, since it's a tree, so you could do 1 read, then 2, then 4, .. etc, for a total of around 7 sequential phases (one for each table in the plot). We haven't got around to doing this yet.

**Comment by mariano54** [comment]
Also another thing to point out is that the responses are returned to the full node as they come out from the drives, so the high time is probably only affecting the slow drive or the slow NAS.

**Comment by coding-horror** [comment]
Yeah, this only started happening as I added multiple NAS devices to the network. With 1 or 2 NASes, it was all fine. Once you get to 5.. not so much, especially if the algorithm picks plots on 6 different devices. At the very least 

- definitely change the log level  to ERROR (or WARN) 
- surface this in the UI for sure, otherwise you're silently "farming".. absolutely nothing. A silent killer.
- tell us WHICH files were picked for proofs so we can look for "problem" devices

thanks @mariano54  !

**Comment by immortalx** [comment]
Is there any benefit of using larger plot files in this case? Less files per directory in his case.

**Comment by keliew** [comment]
I think the problem is inherent in how Windows manages mounted network drives.

If you'd to switch to a Linux-based farmer, it should work better. I know it's not an option for most people, but it'd be a good test to see if that's the cause.

**Comment by kimklai** [comment]
I think lots of people need to know their whole system just too slow to provide the a valid answer in time. So please mark the logs as WARNING as they takes longer a certain threshold.

**Comment by desek** [comment]
This was quite a rough finding since I've been happily farming on my RPi on wifi to a remote storage with proof checking usually between 60-90 seconds.

**Comment by tartley** [comment]
Could some drives be powering down when idle?

**Comment by goldcd** [comment]
Is there any sanity test you can use, without having to restart Chia?

I had thought along the lines of checking plots with "-n 5" (minimum 5 challenges) and if this took longer than 3.5 mins, you might have a problem - except this would indicate I have a very large issue with my very modest plot on my NAS.

But then begs question as to why these check challenges are slower than the regular ones I hope I'm smashing through.

**Comment by coding-horror** [comment]
no, all the NAS devices default to "never sleep drives".

**Comment by bertybassett** [comment]
yeah my 1 Synology NAS is doing the same if I hit it hard with other services such as docker, Plex, etc

**Comment by cccat6** [comment]
> We can't change the consensus algorithm anymore, but we can change the log level, and show the files.
> 
> Please note that looking up qualities for plots passing the filter requires about 7 random reads in a plot, whereas actually looking up a proof requires 64 reads. It might not be feasible on a slow NAS, since these are sequential reads. Furthermore, you need to take into account network latency to propagate your proof and block to the network, so you should be under 5 seconds to reduce risk of losing rewards.
> 
> Actually, the proof of space library does them sequentially, but they could be done in parallel, since it's a tree, so you could do 1 read, then 2, then 4, .. etc, for a total of around 7 sequential phases (one for each table in the plot). We haven't got around to doing this yet.

My disk array is not in the local but distributes in different places. I gather them as a huge disk and connect them with NFS. The latency of this solution for each random seek is between 200-1000ms. The average and the mean is about 400ms.
I wondering whether this is able to farm. 400ms*(7+64)seeks will be 28.4s. Just a little bit below the 30s limitation.
However, this solution support parallel read. If it is seeking 64 different places in the plot at the same time, it will be able to respond to all of them within a single seek time, which is about 400ms.
Is parallel seeking in the todo list and is that work as what I am thinking? When it will be updated?
Thanks a lot!

**Comment by coding-horror** [comment]
Yeah this is a critical issue for the project IMO, since a LOT of people are probably "farming" absolutely nothing due to the 30s harvester timeout, and the logs aren't WARN-ing or ERROR-ing them.. the GUI isn't telling them.. the only way to know this is happening is to intentionally set log level to INFO and scan for 30s or longer in the INFO messages 😱

😭

**Comment by dorkmo** [comment]
i believe the small time window was designed to push tape drives out of the chia eco system. ideally there would be big flashing warnings if youre missing any rewards. hopefully can be addressed in future releases.

**Comment by desek** [comment]
I did a somewhat scalable workaround for this by running multiple harvesters in containers (on the RPi4), each with *ONE* plot directory with 100 plots. This gives arounds 2 seconds for proof checking per container (up to 100 plots).

**Comment by goldcd** [comment]
Had a bit of a panic over this (I've got my plots on a Synology NAS)
Panic compounded when I tried to check my plots - as seemed a fair mechanism to check how quickly they could be challenged. Minimum it allows is "-n 5" (5 checks) so was hoping verifying my plots would be under 5x30 seconds.

So I stopped plotting, enabled INFO and except for a couple of outliers was getting 100 plots done sub-second and all is well.

My assumptions is that check actually runs more challenges and n is the expected number of passes - if check in real world only passes 1/512 times, then maybe the "-n 5" is actually checking 5*512 times (i.e. on average it's expecting 5 passes from 2560 checks) 

Just ran test on my 100 plots and n 5 takes ~ 17 mins  (1020 secs)
So assuming that's 5*512 checks on the set, gives ~0.4 secs for check on all my plots, which seems to be within the bounds of what I see coming back from INFO logging.

**Comment by bertybassett** [comment]
isn't eh simple answer make each Nas its now farmer until chia fix this.

**Comment by qfishpear** [comment]
It's soooooo sad that by "grep eligible" I found nothing there!
(only 1.5TB in my farm, so.... it's just reasonable)

**Comment by Sagittarius** [comment]
You should try "farming on many machines"

**Comment by mariano54** [comment]
You fast ones will still answer in a fast speed, since it's all threaded. It will just display the time of when the slowest finished. 
Most users can farm many plots in under a second, so this is not an issue for most people. But it's worth checking if you're not winning rewards for a while.

**Comment by fiveangle** [comment]
> You fast ones will still answer in a fast speed, since it's all threaded. It will just display the time of when the slowest finished.
> Most users can farm many plots in under a second, so this is not an issue for most people. But it's worth checking if you're not winning rewards for a while.

It seems near criminal that we are conditioned to enable "INFO" level logs and routinely told to ignore the countless "ERROR" level scary sounding messages spamming the logs, yet not a whimper about anything in the logs when a proof challenge request lookup is nearing or fully exhausting some non-presented timeout value that only tribal knowledge or code inspection is aware of.

I watched incredulously as my friends with far fewer plots reached, then far surpassed and doubled over my winnings with 1/2 the plots.  Only once reaching out after the damage is done do we find out with an obtuse, "OF COURSE YOU WON'T WIN WITH LOOKUP DELAYS LIKE THIS" are we made aware of the worst of all actually "errors".

Like I said, that these messages are not flagged as ERROR level or at least WARNING is essentially criminal, especially in the face of the constant stream of non-critical "ERROR" level messages spamming the logs.
